---
title: "FungiExpresZ Evaluation"

output:
       prettydoc::html_pretty:
        toc: TRUE
    # #   highlight: "github"
        fig_caption: TRUE
        # word_document:
        #   toc: TRUE
    # html_document:
    #   theme: "flatly"
    #   toc: TRUE
    #   toc_float: TRUE
      # md_document:
      #   toc: TRUE
      # theme: "flatly"
    
editor_options: 
  chunk_output_type: inline
#runtime: shiny
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.align="center", fig.show="hold")
```


# Data Reference  
***

<div style="text-align: justify"> To evaluate the performance of the FungiExpresZ, we select a data-set from one of the published [study](https://www.nature.com/articles/ncomms11704). Same data can also be obtained from [Supplementary Data 5](https://www.nature.com/articles/ncomms11704#s6) of original publication. The selected data contains 10 RNA-seq samples - 5 from **widtype (WT)** and 5 from **_tetO-HSP90/hsp90Δ_** strain of _Candida albicans_. Each sample has been given in replicates of 3. In this document, we show that, how FungiExpresZ can be used to quickly analyse and visualize given RNA-seq data with or without incorporation of public data. We used log2 of normalized gene expression values (log2(FPKM)) for all analysis given below. </div>

## Sample name and description  
 
|Sr.| Strain type |    Sample name          |	Interpretation                                                                      |
| - |:--------:   |:------------:           |----------------------------------------------------------------------------------------------|
| 1 | Wild type   |wt_no_dox_no_HS          |	Wild type, absence of doxycycline, no heat shock                                    |
| 2 | Wild type   |wt_dox_no_HS             |	Wild type, presence of  20 μg/ml doxycycline, no heat shock                         |
| 3 | Wild type   |wt_dox_10_HS             |	Wild type, presence of  20 μg/ml doxycycline, 10 min post 30–42°C heat shock        |
| 4 | Wild type   |wt_dox_30_HS             |	Wild type, presence of  20 μg/ml doxycycline, 30 min post 30–42°C heat shock        |
| 5 | Wild type   |wt_dox_60_HS             |	Wild type, presence of  20 μg/ml doxycycline, 60 min  post 30–42°C heat shock       |
| 6 |_tetO-HSP90/hsp90Δ_|tet90_no_dox_no_HS	| tetO-HSP90/hsp90Δ, absence of doxycycline, no heat shock                            |
| 7 |_tetO-HSP90/hsp90Δ_|tet90_dox_no_HS    | tetO-HSP90/hsp90Δ, presence of  20μg/ml doxycycline, no heat shock                  |
| 8 |_tetO-HSP90/hsp90Δ_|tet90_dox_10_HS    | tetO-HSP90/hsp90Δ, presence of  20μg/ml doxycycline, 10 min post 30–42°C heat shock |
| 9 |_tetO-HSP90/hsp90Δ_|tet90_dox_30_HS    | tetO-HSP90/hsp90Δ, presence of  20μg/ml doxycycline, 30 min post 30–42°C heat shock |
| 10|_tetO-HSP90/hsp90Δ_|tet90_dox_60_HS    | tetO-HSP90/hsp90Δ, presence of  20μg/ml doxycycline, 60 min  post 30–42°C heat shock|
  

# Step 1: Pair-wise correlation between replicates  
***

All pair of replicates in both WT and _tetO-HSP90/hsp90Δ_ given in Fig: 1.1 and Fig: 1.2 respectively, show Pearson correlation value than 0.9 suggesting very high reproducibility of the biological experiments. 

## Wild type  
```{r , out.width = "50%" , echo=FALSE, fig.cap="Fig 1.1: Multiple pair wise scatter plots depicting correlation between wild type replicates"} 
replicates_corr_plots <-list.files(path = "./eval_plots/01/", pattern = "\\d.*wt.*multi_scatter_plot.png", full.names = TRUE)
knitr::include_graphics(replicates_corr_plots)
```

## _tetO-HSP90/hsp90Δ_  

```{r , out.width = "50%" , echo=FALSE, fig.cap="Fig 1.2: Multiple pair wise scatter plots depicting correlation between _tetO-HSP90/hsp90Δ_ replicates"} 

replicates_corr_plots <-list.files("./eval_plots/01/", pattern = "\\d.*del.*multi_scatter_plot.png", full.names = TRUE)
knitr::include_graphics(replicates_corr_plots)
```



# Step 2: Pair-wise correlation within WT and _tetO-HSP90/hsp90Δ_ strain samples 
***

<div style="text-align: justify">  Correlation between samples of identical strain will allow us to quickly identify if any of the experiments have dramatic effect of heat or doxycycline. Fig: 2.1 and Fig: 2.2  show correlation within **WT** and **_tetO-HSP90/hsp90Δ_** respectively.  

In WT strain (Fig: 2.1), Pearson correlation value between **wt_no_dox_no_hs** and **wt_dox_no_hs** is more than 0.97 suggesting very low or no effect of doxycycline in WT in absence of heat-shock. However, the Pearson correlation between **wt_dox_no_hs** and **wt_dox_30_hs**, and **wt_dox_no_hs** and **wt_dox_60_hs** are 0.90 and 0.91 respectively suggesting that, in presence of doxycycline, 30 and 60 minutes post heat-shock, genes expression seems altered compared to no heat-shock.

In **_tetO-HSP90/hsp90Δ_** strain(Fig: 2.2),  correlation between **tet90_no_dox_no_hs** and **tet90_dox_no_hs** is 0.96 suggesting that, presence of doxycycline does not bring dramatic change of gene expression. However, correlation between **tet90_dox_no_hs** and **tet90_dox_10_hs**, and **tet90_dox_no_hs** and **tet90_dox_60_hs** are 0.91 and 0.89 respectively, suggesting that, in presence of doxycycline, 10 and 60 minutes post heat-shock, genes expression found altered compared to no heat-shock.  </div>

## Wild type  

```{r echo=FALSE,  out.width="90%" , fig.cap= "Fig 2.1: Multiple pair wise scatter plots depicting correlation within  WT strain samples"}
within_strain_corr_plots_wt <-c("./eval_plots/01/14_corr_within_wt_samples.png")
knitr::include_graphics(within_strain_corr_plots_wt)
```

## _tetO-HSP90/hsp90Δ_  

```{r echo=FALSE,  out.width="90%" , fig.cap= "Fig 2.2: Multiple pair wise scatter plots depicting correlation within _tetO-HSP90/hsp90Δ_ strain samples"}
within_strain_corr_plots_del <-c("./eval_plots/01/13_corr_within_del_samples.png")
knitr::include_graphics(within_strain_corr_plots_del)
```

# Step 3: Pair-wise correlation between WT and _tetO-HSP90/hsp90Δ_ strain samples  
***

```{r  , out.width = "50%" , echo=FALSE, fig.cap="Fig 3.1: Multiple pair wise scatter plots depicting correlation between  WT and _tetO-HSP90/hsp90Δ_"} 
wt_vs_del_corr_plots <-list.files("./eval_plots/01/", pattern = "\\d+_wt_vs_del.*.png", full.names = TRUE)
knitr::include_graphics(wt_vs_del_corr_plots)
```
To check the effect of HSP90 depletion, we compared **WT** vs **_tetO-HSP90/hsp90Δ_** in all combinations. 


# Step 4: Identify similarity and differences between samples through Principal component analysis (PCA) 
***
<div style="text-align: justify"> PCA plots of **WT** and **_tetO-HSP90/hsp90Δ_** are shown in the Fig: 4.1 and Fig: 4.2 respectively. Both **WT** and **_tetO-HSP90/hsp90Δ_** PCA plots show that, replicates (A,B and C) from each sample, cluster together, suggesting reproducibility of biological experiments. In the **WT** PCA plot, samples with doxycycline 30 and 60 minutes post heat-shock (WT_dox_30_HS , WT_dox_60_HS) are placed far a part compare to the samples without heat-shock in presence and absence of doxycycline (WT_no_dox_no_HS, WT_dox_no_HS). While PCA plot of **_tetO-HSP90/hsp90Δ_** shows that, sample with doxycycline 30 min post heat-shock (tet90_dox_30_HS) is more similar to no heat-shock with doxycycline (tet90_dox_no_HS). Overall, our PCA plot shows similarity and differences of gene expression profile within **WT** and **_tetO-HSP90/hsp90Δ_** of _Candida albicans_.  <div>

## Wild type

```{r , out.width = "90%" , echo=FALSE , fig.cap="Fig 4.1: Principle component analysis within wild type samples"} 
pca_plot_wt <-list.files("./eval_plots/01/", pattern = "\\d.*wt.*pca_plot.png", full.names = TRUE)
knitr::include_graphics(pca_plot_wt)
```


## _tetO-HSP90/hsp90Δ_  

```{r , out.width = "90%" , echo=FALSE , fig.cap="Fig 4.2: Principle component analysis within _tetO-HSP90/hsp90Δ_ samples"} 
pca_plot_del <-list.files("./eval_plots/01/", pattern = "\\d.*del.*pca_plot.png", full.names = TRUE)
knitr::include_graphics(pca_plot_del)
```


# Step 5: Identifying genes of interest from scatter plot  
***

<div style="text-align: justify"> Each dot represents gene expression value(FPKM) of a gene under two conditions. Any point deviating from population can be seen as an outlier. Since these genes don’t follow normal distribution they can potentially interesting genes biologically. However, without prior programming knowledge it is challenging to track these biologically significant genes. Here, we setup one stop solution to analyse huge genomics data under multiple condition for any organism. FungiExpresZ provides a feature to select set of genes readily from the plot through mouse drag function and swiftly displays their annotations along in downloadable format.  </div>

As an example, we show correlation between **tet90_dox_no_hs** and **tet90_dox_30_hs**. Mouse-over functionality readily displays annotations of the selected genes in tabular format. The annotations are displayed in the table below. For instance, HSP12 a heat shock protein shows induced expression at 10min post heat shock compared to no heat condition. FungiExpressz allows user to download data table on click for future analysis.

```{r echo=FALSE , out.width="100%" , fig.cap="Fig 5.1: A Sctter plot showing genes selected with mouse drag function"}
corr_plot_pair3 <- c("./eval_plots/01/15_corr_plot_genes_highlight_03.png")
knitr::include_graphics(corr_plot_pair3)
```

```{r echo=FALSE , out.width="80%" , fig.cap="Fig 5.2: A table showing genes' annotations , which are highlighted in Fig: 5.1"}
corr_plot_pair3_table <- c("./eval_plots/01/16_corr_plot_genes_highlight_table_03.png")
knitr::include_graphics(corr_plot_pair3_table)
```

# Step 6: Detection of highly variable genes across different conditions
***

## Derive gene clusters through heatmap  

The heatmap below shows the top 1000 most variable genes across the conditions, grouped into four different clusters. FungiExpresZ allows you to set the number of genes to be displayed and number of clusters to which genes are grouped in. 

```{r echo=FALSE , out.width="50%",fig.cap="Fig 6.1: Heatmap showing top 1000 highly variable genes across conditions grouped into four different clusters"}
top_var_genes_hm <- c("./eval_plots/01/23_wt_hsp_top_1000.png")
knitr::include_graphics(top_var_genes_hm)
```


## Display gene clusters through joy plot  

```{r , echo=FALSE , out.width="50%", fig.cap="Fig 6.2: Joy plot depicts all clusters from heatmap" }
joy_plot_by_cluster <-list.files("./eval_plots/01/", pattern = "\\d+.*clust.*joy_plot.png", full.names = TRUE)
knitr::include_graphics(joy_plot_by_cluster)
```
## Gene expression pattern of each cluster

### Cluster 1  
It contains the genes, which show **early transcriptional response** upon depletion of **_tetO-HSP90/hsp90Δ_** compare to **WT**, as level of gene expression post 30 minutes heat-shock in **WT** is similar to post 10 minutes heat-shock in **_tetO-HSP90/hsp90Δ_** strain.

### Cluster 2  
It contains the genes, which show **delayed transcriptional response** upon depletion of **_tetO-HSP90/hsp90Δ_** compare to **WT**, as level of gene expression post 30 minutes heat-shock in **_tetO-HSP90/hsp90Δ_** is similar to post 10 minutes heat-shock in **WT** strain. 

### Cluster 3  
It contains the genes, which show **delayed transcriptional response** upon depletion of **_tetO-HSP90/hsp90Δ_** compare to **WT**, as level of gene expression post 60 minutes heat-shock in **_tetO-HSP90/hsp90Δ_** is similar to post 30 and 60 minutes heat-shock in **WT** strain. 

### Cluster 4  
It contains the genes, which show **longer transcriptional response** upon depletion of **_tetO-HSP90/hsp90Δ_** compare to **WT**, as level of gene expression post 30 and 60 minutes heat-shock in **_tetO-HSP90/hsp90Δ_** is similar to post 60 minutes heat-shock in **WT** strain. 

