---
title: "Newz and Updates"
author: "Chirag Parsania"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output:
     prettydoc::html_pretty:
      number_sections: FALSE
      toc: TRUE
      fig_caption: TRUE
      toc_depth: 4
editor_options: 
  chunk_output_type: inline

---


<style>
body {
text-align: justify}
</style>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
app_name <-  "FungiExpresZ"
library(tidyverse)
library(knitr)
library(kableExtra)
```

### **About this page** 

This page is all about notifing  users regarding new features, bug fixing, any other important changes or announcements made on `r app_name` over the time. 

***

### **`r lubridate::date(x = "2019/10/14")`**
#### New logo introduced

`r app_name` has its identity now. Introducing new logo ðŸŽ‰ðŸŽ‰ðŸŽ‰ðŸŽ‰ 

  <a href='https://raw.githubusercontent.com/cparsania/FungiExpresZ/master/inst/app/www/hex_sticker.png' target="_blank"><img src='https://raw.githubusercontent.com/cparsania/FungiExpresZ/master/inst/app/www/hex_sticker.png'  height="300" /></a>
  
***




### **`r lubridate::date(x = "2019/10/10")`**
#### Available as a docker image 

`r app_name` is now available as a docker image and can be run as a container on any local computer where docker hub is installed. Follow instructions given on [Github](https://github.com/cparsania/FungiExpresZ){target="_blank"} to install and run locally. 

***

### **`r lubridate::date(x = "2019/10/08")`**
#### Available as an R package (v0.0.1)

`r app_name` can be installed as an R package and run on local computer. For local installation, please follow instruction on [Github](https://github.com/cparsania/FungiExpresZ#install){target="_blank"}

***

### **`r lubridate::date(x = "2019/09/23")`**
#### New SRA data added for **_Aspergillus fumigatus_** and **_Aspergillus niger_**  

Number of samples increased to 242 from 133 and 253 from 28 for **_Aspergillus fumigatus_** and **_Aspergillus niger_**  respectively. 

***


### **`r lubridate::date(x = "2019/09/18")`**
#### Download data  
  + `r app_name` allows user to download gene ontology data and gene expression matrix in.txt file format. Data can be downloaded under the tab **Downloads**. 

***

### **`r lubridate::date(x = "2019/09/16")`**
#### Improved line plot  
  + Performance of line plot is improved. 
  + Along with gene clusters, individual genes can also be plotted.
  + Color and facet by gene groups can be done.
  + Line size and transparancy can be changed. 

***
### **`r lubridate::date(x = "2019/09/14")`**
#### Video tutorials
  + `r app_name`  tutorial videos have been given under tab **Tutorial**. User can also watch videos on [YouTube channel](https://www.youtube.com/watch?v=B4hwnLijEB8&list=PLO5JyZm5J6sw4nWDBtyLj813BfhTYfBzz){target="_blank"}.
  
***
### **`r lubridate::date(x = "2019/09/09")`**
#### Display active groups  
  + User can see now active gene groups and active sample groups any time while accessing the app. Groups can be seen under the section of sidepanel **View active groups**.
  
***
### **`r lubridate::date(x = "2019/08/30")`**
#### New GO visualizations: UpSet plot and Bar plot   
  + Two more GO visualizations added : **1)**  UpSet plot and **2)** Bar plot. With addition of these two GO plots, `r app_name` has now total 6 different GO visualizations. GO visualizations can be done for the genes selected from scatter plot or gene clusters selected from line plot and heatmap. 
  
***
### **`r lubridate::date(x = "2019/08/27")`**
#### Use NCBI BioProject as sample groups
  + While using the SRA data given in the `r app_name`, user can use NCBI BioProjects as sample groups. Using NCBI BioProjects as sample groups allow user to compre between SRA samples of multiple studies. To use NCBI BioProjects as sample groups follow this  **Assign groups** --> **Samples**  --> **Group by BioProjects (NCBI)**.
  
*** 
### **`r lubridate::date(x = "2019/08/09")`**
#### Access SRA sample information for heatmap column clusters
  + Most obvious questions to ask once the heatmap created are **1)** what are the samples under column clusters and **2)** what are the genes under the row clusters. `r app_name` has already provided GO analysis feature for the row clusters. Now, the new feature allow user to get sample informations for a specified  column cluster. For a selected column cluster `r app_name` displayes below informations.

```{r , out.width="50%" , echo =FALSE}
  knitr::kable(tibble::tibble("Heatmap column label", "NCBI Bio project ID", "Library Name", "Strain", "Genotype", "Study title", "Study abstract"), col.names = NULL) %>% 
  kableExtra::kable_styling(bootstrap_options = "striped", full_width = T, position = "center")
```
  
#### Visualize heatmap column cluster sample information in the word cloud plot
  + Heatmap is the powerful way to visualize the gene expression data. In general, columns represents different samples and rows represents genes. Implying various clustering methods on heatmap rows and column can identify clusters of similiar expression profiles. For ease of understanding samples under given column cluster, `r app_name` provides word cloud visualization. For a given column cluster, it collects unique set of study abstract and perform word enrichment using text mining. The enriched words are displayed in form of word cloud plot. Word cloud feature can be accessed once the heatmap created using SRA samples.

***
### **`r lubridate::date(x = "2019/08/01")`**

#### Newz and update page introduced
  + First time **Newz and Update** page introduced. Form now onwards, updates / new feature / bug fixing etc. will be reported on this page. 

***







  
  
  
